core/src/main/java/org/apache/calcite/rex/RexBuilder.java

```diff
@@ -699,22 +699,23 @@ boolean canRemoveCastFromLiteral(RelDataType toType, @Nullable Comparable value,
 
     if (SqlTypeName.INT_TYPES.contains(sqlType)) {
       final BigDecimal decimalValue = (BigDecimal) value;
-      final int s = decimalValue.scale();
-      if (s != 0) {
+      try {
+        // Will throw if the value does not fit in a long
+        long l = decimalValue.longValueExact();
```

Comment:
check if the decimal value will fit in the target type

Reply:
Should these checks factor in unsigned types?

Reply:
There are no unsigned types in SQL.